examples/threading/threading.c

```diff
@@ -0,0 +1,91 @@
+#include "threading.h"
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+
+// Optional: use these functions to add debug or error prints to your application
+#define DEBUG_LOG(msg,...)
+//#define DEBUG_LOG(msg,...) printf("threading: " msg "\n" , ##__VA_ARGS__)
+#define ERROR_LOG(msg,...) printf("threading ERROR: " msg "\n" , ##__VA_ARGS__)
+
+
+// convert msec to usec
+#define MILLISECOND_TO_USEC    (1000)
+
+
+void* threadfunc(void* thread_param)
+{
+
+    // TODO: wait, obtain mutex, wait, release mutex as described by thread_data structure
+    // hint: use a cast like the one below to obtain thread arguments from your parameter
+    //struct thread_data* thread_func_args = (struct thread_data *) thread_param;
+    int ret;
+    bool local_success = false;
+    struct thread_data* args = (struct thread_data *) thread_param;
+
+    // wait to start thread
+    usleep((args->wait_to_obtain_ms)*MILLISECOND_TO_USEC);
+
+    // call lock on mutex
+    ret = pthread_mutex_lock(args->mutex);
+    if(ret != 0) {
+        ERROR_LOG("Error in acquiring lock in TID %lu. %s", pthread_self(), strerror(errno));
```

Comment:
Even when it errors out, it is executing critical section. You need to either free the thread params memory and return NULL or simply return the thread params which will be freed in test code